src/Artifacts.jl

```diff
@@ -0,0 +1,594 @@
+        end
+
+        # Give the people what they want
+        return artifact_hash
+    finally
+        # Always attempt to cleanup
+        rm(temp_dir; recursive=true, force=true)
+    end
+end
+
+"""
+    artifact_path(hash::SHA1)
+
+Given an artifact (identified by SHA1 git tree hash), return its installation path.
+"""
+artifact_path(hash::SHA1) = artifacts_dir(bytes2hex(hash.bytes))
+
+"""
+    artifact_exists(hash::SHA1)
+
+Returns whether or not the given artifact (identified by its sha1 git tree hash) exists
+on-disk.  If it does not, 
+"""
+artifact_exists(hash::SHA1) = isdir(artifact_path(hash))
+
+"""
+    remove_artifact(hash::SHA1)
+
+Removes the given artifact (identified by its SHA1 git tree hash) from disk.
+"""
+remove_artifact(hash::SHA1) = rm(artifact_path(hash); recursive=true, force=true)
+
+"""
+    verify_artifact(hash::SHA1)
+
+Verifies that the given artifact (identified by its SHA1 git tree hash) is installed on-
+disk, and retains its integrity. 
+"""
+function verify_artifact(hash::SHA1)
+    if !artifact_exists(hash)
+        return false
+    end
+
+    return hash.bytes == tree_hash(artifact_path(hash))
+end
+
+"""
+    archive_artifact(hash::SHA1, tarball_path::String)
+
+Archive an artifact into a tarball stored at `tarball_path`, returns the SHA256
+of the resultant tarball.  Throws an error if the artifact does not exist.
+"""
+function archive_artifact(hash::SHA1, tarball_path::String)
+    if !artifact_exists(hash)
+        error("Unable to archive artifact $(bytes2hex(hash.bytes)): does not exist!")
+    end
+
+    # Package it up
+    package(artifact_path(hash), tarball_path)
+
+    # Calculate its sha256 and return that
+    return open(tarball_path, "r") do io
+        return bytes2hex(sha256(io))
+    end
+end
+
+
+"""
+    unpack_platform(entry::Dict, name::String, artifacts_toml::String)
+
+Given an `entry` for the artifact named `name`, located within the file `artifacts_toml`,
+returns the `Platform` object that this entry specifies.  Returns `nothing` on error.
+"""
+function unpack_platform(entry::Dict, name::String, artifacts_toml::String)
+    if !haskey(entry, "os")
+        @warn("Invalid Artifacts.toml at '$(artifacts_toml)': platform-specific artifact entry '$name' missing 'os' key")
+        return nothing
+    end
+
+    if !haskey(entry, "arch")
+        @warn("Invalid Artfiact.toml at '$(artifacts_toml)': platform-specific artifact entrty '$name' missing 'arch' key")
+        return nothing
+    end
+
+    # Helpers to pull out `Symbol`s and `VersionNumber`s while preserving `nothing`.
+    nosym(x::Nothing) = nothing
+    nosym(x) = Symbol(lowercase(x))
+    nover(x::Nothing) = nothing
+    nover(x) = VersionNumber(x)
+
+    # First, extract OS; we need to build a mapping here
+    os_map = Dict(
+        "windows" => Windows,
+        "macos" => MacOS,
+        "freebsd" => FreeBSD,
+        "linux" => Linux,
+    )
+    P = get(os_map, lowercase(entry["os"]), UnknownPlatform)
```

Comment:
I wonder if we should define:
`parse(::Type{Type{Platform}}, str)`

Reply:
https://github.com/JuliaLang/Pkg.jl/blob/ee09076490a3a2e2445044cbb11eb4033b0ac1aa/src/BinaryPlatforms.jl#L472-L580